### PR TITLE
New rule: disallowUnusedVariables

### PR DESCRIPTION
**Pull request by schempy**
disallowUnusedVariables: warn/auto-fix unused variables defined with var, let or const.

Fixes: #2076

### CODE REVIEW

**Comment by schempy** [comment]
The Travis CI build shows all tests passing. I noticed the following message from the node v0.10.44 build:
```
Bad response: 422 {"message":"Couldn't find a repository matching this job.","error":true}
```

Not sure if that is from coveralls. I did a rebase from the current master branch prior to the pull request.

**Comment by markelog** [comment]
Wow, if everything works, that would be an impressive one!

/cc @mdevils 

> Bad response: 422 {"message":"Couldn't find a repository matching this job.","error":true}

You probably force-push at some point, which what happens, i.e. totally okay.

File: lib/rules/disallow-unused-variables.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * ```
+ *
+ * ##### Invalid
+ *
+ * ```js
+ * var x =1;
+ *
+ * function getX() {
+ *     var x=1;
+ *     return x;
```
Comment (markelog) [inline lib/rules/disallow-unused-variables.js]: It seems in this case, variable is used?

**Comment by schempy** [comment]
Thanks for catching the invalid example. I updated.

**Comment by markelog** [comment]
@schempy you are the man! Thanks

**Comment by jdalton** [comment]
This is having issues:
```bash
disallowUnusedVariables: Variable `object` is not used at lodash.js :
 14622 |     */
 14623 |    var method = rest(function(path, args) {
 14624 |      return function(object) {
---------------------------------^
 14625 |        return baseInvoke(object, path, args);
 14626 |      };
```

You can see the var is used

**Comment by markelog** [comment]
That was fast! @schempy could you check it out?

**Comment by mdevils** [comment]
Thanks, will fix it soon.

**Comment by mdevils** [comment]
Fixed in `3.0.1`. Please check it out.

**Comment by jdalton** [comment]
It works! Updated dep in Lodash :D

**Pull request closed without merging**